### PR TITLE
SQLStore: Ensure that sessions are always closed

### PR DESCRIPTION
**Pull request by papagian**
**What this PR does / why we need it**:
- It removes `NewSession()` from [db.DB](https://github.com/grafana/grafana/blob/ccd4a6c4a3bfbf337cae72a8313a0b3e50c291cb/pkg/services/sqlstore/db/db.go#L11) and [sqlstore.Store](https://github.com/grafana/grafana/blob/ccd4a6c4a3bfbf337cae72a8313a0b3e50c291cb/pkg/services/sqlstore/store.go#L12) interfaces (in favour of `WithDbSession()`)
- It also introduces [WithDbSessionForceNewSession()](https://github.com/grafana/grafana/blob/ccd4a6c4a3bfbf337cae72a8313a0b3e50c291cb/pkg/services/sqlstore/session.go#L63) that always creates a new session (as opposed to the [WithDbSession()](https://github.com/grafana/grafana/blob/ccd4a6c4a3bfbf337cae72a8313a0b3e50c291cb/pkg/services/sqlstore/session.go#L58)  (currently not used) that loads the session from the context - if one exists) 

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

### CODE REVIEW

File: pkg/services/sqlstore/session.go [hunk 1]
```diff
@@ -55,14 +48,24 @@ func startSessionOrUseExisting(ctx context.Context, engine *xorm.Engine, beginTr
 	}
 
 	newSess.Session = newSess.Session.Context(ctx)
+
 	return newSess, true, nil
 }
 
-// WithDbSession calls the callback with a session.
+// WithDbSession calls the callback with the session in the context (if exists).
```
Comment (mildwonkey) [inline pkg/services/sqlstore/session.go:55]: thank you for the comments <3

File: pkg/services/sqlstore/session.go [hunk 2]
```diff
@@ -55,14 +48,24 @@ func startSessionOrUseExisting(ctx context.Context, engine *xorm.Engine, beginTr
...
+// Otherwise it creates a new one that is closed upon completion.
+// A session is stored in the context if sqlstore.InTransaction() has been been previously called with the same context (and it's not committed/rolledback yet).
 func (ss *SQLStore) WithDbSession(ctx context.Context, callback DBTransactionFunc) error {
 	return withDbSession(ctx, ss.engine, callback)
 }
 
+// WithDbSession calls the callback with a new session that is closed upon completion.
```
Comment (mildwonkey) [inline pkg/services/sqlstore/session.go:62]: nit: update method name in the comment please

File: pkg/services/sqlstore/session.go [hunk 3]
```diff
@@ -55,14 +48,24 @@ func startSessionOrUseExisting(ctx context.Context, engine *xorm.Engine, beginTr
...
+func (ss *SQLStore) WithDbSessionForceNewSession(ctx context.Context, callback DBTransactionFunc) error {
```
Comment (mildwonkey) [inline pkg/services/sqlstore/session.go:63]: nit: this method name is a bit confusing to me. I don't think this name is perfect either, but maybe something closer to `ForceNewSession()`?

File: pkg/infra/remotecache/database_storage.go [hunk 4]
```diff
@@ -54,36 +54,40 @@ func (dc *databaseCache) internalRunGC() {
 
 func (dc *databaseCache) Get(ctx context.Context, key string) (interface{}, error) {
 	cacheHit := CacheData{}
-	session := dc.SQLStore.NewSession(ctx)
-	defer session.Close()
 
-	exist, err := session.Where("cache_key= ?", key).Get(&cacheHit)
+	var r interface{}
```
Comment (mildwonkey) [inline pkg/infra/remotecache/database_storage.go:58]: the tiniest of nitpics: I believe if you initialize  a `cachedItem` here , you can use that on line 82 and skip initializing a second variable and continue to return `item.Val` at the end.

**Review by mildwonkey — COMMENTED** [review]
This looks great @papagian and I really appreciate the extra comments you added. 
I left some *very tiny* and *entirely ignorable* nitpicks, but overall this looks reasonable to me. (someone else should review this too; I'm not familiar enough with the inner workings of our sqlstore)

File: pkg/services/sqlstore/session.go (hunk 3 above)
Comment (papagian) [reply pkg/services/sqlstore/session.go:63]: WDYT about `WithNewDbSession()`?

Comment (mildwonkey) [reply pkg/services/sqlstore/session.go:63]: I like it as long you think it conveys the necessary information!

**Review by mildwonkey — APPROVED** [review]
Looks great - thanks for walking me through some of this in slack!

**Review by ying-jeanne — APPROVED** [review]
AMAZA

**Review by joeblubaugh — APPROVED** [review]
Alerting code all looks great.

**Pull request merged**